File: python/tests/test_cache_concurrency.py
"""TC9 — Cache thread-safety.

The TTL cache is shared mutable state on the client; the caching suite only
exercises it single-threaded. These fire concurrent reads and concurrent
mutations at one cached client and assert no corruption/exception and a
consistent result.

The generated Python Airtable has process-global/class-level shared state
(``set_airtable_config`` plus a class-level ``_tables`` cache). If concurrent
access corrupts that shared state or raises, that is a real finding.

Mirrors csharp/tests/TestCacheConcurrency.cs.
"""

import uuid
from concurrent.futures import ThreadPoolExecutor

import pytest

from output import Airtable, PrimaryModel


@pytest.fixture(scope="module")
def airtable_cached():
    return Airtable(cache_seconds=60)


def _primary_key(scenario: str) -> str:
    return f"CacheConc {scenario} {uuid.uuid4().hex[:8]}"


def _try_delete(airtable: Airtable, record_id: str | None) -> None:
    if not record_id:
        return
    try:
        airtable.primary.delete(record_id=record_id)
    except Exception:
        pass


def test_concurrent_gets_of_the_same_record_are_consistent(airtable_cached: Airtable):
    key = _primary_key("Reads")
    model = PrimaryModel()
    model.primary_key = key
    created = airtable_cached.primary.create(model)
    record_id = created.id

    def get_one() -> PrimaryModel:
        return airtable_cached.primary.get(record_id)

    try:
        # 25 concurrent gets race to populate/read the shared cache.
        with ThreadPoolExecutor(max_workers=25) as pool:
            futures = [pool.submit(get_one) for _ in range(25)]
            results = [f.result() for f in futures]
        assert all(m.primary_key == key for m in results)
        assert all(m.id == record_id for m in results)
    finally:
        _try_delete(airtable_cached, record_id)


def test_concurrent_reads_and_mutations_do_not_corrupt_the_cache(airtable_cached: Airtable):
    key = _primary_key("Mix")
    model = PrimaryModel()
    model.primary_key = key
    created = airtable_cached.primary.create(model)
    record_id = created.id

    def get_one() -> PrimaryModel:
        return airtable_cached.primary.get(record_id)

    try:
        # Interleave reads (populate cache), per-table invalidations, and
        # cascade (invalidate-all) invalidations concurrently.
        with ThreadPoolExecutor(max_workers=16) as pool:
            futures = []
            for _ in range(15):
                futures.append(pool.submit(get_one))
                futures.append(pool.submit(airtable_cached.primary.invalidate_cache))
                futures.append(pool.submit(airtable_cached.invalidate_cache))
            for f in futures:
                f.result()

        # The client is still usable and returns the correct value afterward.
        fetched = get_one()
        assert fetched.primary_key == key
    finally:
        _try_delete(airtable_cached, record_id)

File: cpp/tests/test_cache_concurrency.cpp
// Cache thread-safety (C# TestCacheConcurrency parity, 2 cases). The TTL cache
// is shared mutable state on the client; the caching suite only exercises it
// single-threaded. These fire concurrent reads and concurrent mutations at one
// cached client and assert no corruption/exception and a consistent result.
// (Run the cpp_tests_tsan variant to surface unsynchronized access.)
#include <catch2/catch_test_macros.hpp>

#include <exception>
#include <string>
#include <thread>
#include <vector>

#include "test_setup.hpp"

using namespace myairtable;
using myairtable_tests::make_airtable;
using myairtable_tests::primary_key;

namespace {

Airtable cached() {
    return make_airtable(60.0);
}

void best_effort_delete(Airtable& at, const std::string& id) {
    if (id.empty()) {
        return;
    }
    try {
        at.primary().delete_one(id);
    } catch (const AirtableException&) {
    }
}

/// Join every thread, then rethrow the first captured exception (if any).
void join_and_rethrow(std::vector<std::thread>& threads, std::vector<std::exception_ptr>& errors) {
    for (auto& t : threads) {
        t.join();
    }
    for (const auto& error : errors) {
        if (error) {
            std::rethrow_exception(error);
        }
    }
}

} // namespace

TEST_CASE("cache concurrency: concurrent gets of the same record are consistent",
          "[cache][cache-concurrency]") {
    auto at = cached();
    const auto key = primary_key("CacheConc", "Reads");
    auto created = at.primary().create_one(PrimaryModel{.primary_key = key});
    const auto id = *created.id;
    try {
        // 25 concurrent gets race to populate/read the shared cache.
        constexpr int kReaders = 25;
        std::vector<PrimaryModel> results(kReaders);
        std::vector<std::exception_ptr> errors(kReaders);
        std::vector<std::thread> threads;
        threads.reserve(kReaders);
        for (int i = 0; i < kReaders; ++i) {
            threads.emplace_back([&, i] {
                try {
                    results[i] = at.primary().get_one(id);
                } catch (...) {
                    errors[i] = std::current_exception();
                }
            });
        }
        join_and_rethrow(threads, errors);

        for (const auto& model : results) {
            REQUIRE(model.primary_key == key);
            REQUIRE(model.id == id);
        }
    } catch (...) {
        best_effort_delete(at, id);
        throw;
    }
    best_effort_delete(at, id);
}

TEST_CASE("cache concurrency: concurrent reads and mutations do not corrupt the cache",
          "[cache][cache-concurrency]") {
    auto at = cached();
    const auto key = primary_key("CacheConc", "Mix");
    auto created = at.primary().create_one(PrimaryModel{.primary_key = key});
    const auto id = *created.id;
    try {
        // Interleave reads (populate cache), manual invalidations, and
        // re-reads concurrently.
        constexpr int kIterations = 15;
        std::vector<std::exception_ptr> errors(kIterations * 3);
        std::vector<std::thread> threads;
        threads.reserve(kIterations * 3);
        for (int i = 0; i < kIterations; ++i) {
            threads.emplace_back([&, slot = i * 3] {
                try {
                    at.primary().get_one(id);
                } catch (...) {
                    errors[slot] = std::current_exception();
                }
            });
            threads.emplace_back([&, slot = i * 3 + 1] {
                try {
                    at.client()->invalidate_cache(std::string(PrimaryModel::kTableId));
                } catch (...) {
                    errors[slot] = std::current_exception();
                }
            });
            threads.emplace_back([&, slot = i * 3 + 2] {
                try {
                    at.invalidate_all_caches();
                } catch (...) {
                    errors[slot] = std::current_exception();
                }
            });
        }
        join_and_rethrow(threads, errors);

        // The client is still usable and returns the correct value afterward.
        auto fetched = at.primary().get_one(id);
        REQUIRE(fetched.primary_key == key);
    } catch (...) {
        best_effort_delete(at, id);
        throw;
    }
    best_effort_delete(at, id);
}
